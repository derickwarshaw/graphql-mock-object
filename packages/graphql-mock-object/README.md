# graphql-mock-object

> Prototype your UI with GraphQL `mock { ... }` objects before defining complex types & resolvers.

GraphQL works best as a _backend for the frontend (BFF)_ that matches the structure of your UI. `graphql-mock-object` makes it simple for your UI to request data however you'd like. Once you're done prototyping, you can replace mocks with _real_ resolvers and return _real_ data.

### Installation

I recommend saving into `devDependencies`, as prototyping is typically done in `development`, not `production`.

```
yarn add --dev graphql-mock-object
# ...or...
npm install --save-dev graphql-mock-object
```

Then add it to your application:

```js
import { graphiqlExpress, graphqlExpress } from "apollo-server-express"
import bodyParser from "body-parser"
import express from "express"
import { makeExecutableSchema } from "graphql-tools"

// ðŸ‘‡ All the dependencies we need
import { mock, MockObject, typeDefs } from "graphql-mock-object"

const schema = makeExecutableSchema({
  typeDefs: [
    ...typeDefs, // ðŸ‘ˆ All the mock types we're dependent on
    `type Query { version: String }`, // ðŸ‘ˆ Your existing `Query`
    `extend type Query { mock: MockObject! }`, // ðŸ‘ˆ Add `mock` to `Query`
  ],
  resolvers: {
    MockObject, // ðŸ‘ˆ This resolves all mock properties
    Query: {
      mock, // ðŸ‘ˆ This is needed to query `mock`
      version() {
        return require("../../../package.json").version
      },
    },
  },
})

// The rest of your app...
export const app = express()
  .get("/", (req, res) => res.redirect("/graphiql"))
  .use("/graphql", bodyParser.json(), graphqlExpress({ schema }))
  .use("/graphiql", graphiqlExpress({ endpointURL: "/graphql" }))
  .listen(3000, function(err) {
    if (err) {
      console.error(err)
      return
    }

    console.log("Listening at http://localhost:3000")
  })
```

### Author

[Eric Clemmons](https://github.com/ericclemmons/)
([@ericclemmons](https://twitter.com/ericclemmons))
