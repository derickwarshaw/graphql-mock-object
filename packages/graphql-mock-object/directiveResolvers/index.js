export async function fake(next, src, args, context) {
  const value = await next()

  console.log({ src, args, value })

  return value
}
