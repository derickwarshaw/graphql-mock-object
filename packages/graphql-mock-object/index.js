export { directiveResolvers } from "./directResolvers"
export { MockObject, resolvers } from "./resolvers"
export { typeDefs } from "./typeDefs"

export function mock() {
  return {}
}
