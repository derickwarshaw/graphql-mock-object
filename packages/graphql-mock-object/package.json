{
  "author": "Eric Clemmons <eric@smarterspam.com>",
  "bugs": "https://github.com/ericclemmons/graphql-mock-object/issues",
  "dependencies": {},
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "name": "graphql-mock-object",
  "repository": "https://github.com/ericclemmons/graphql-mock-object",
  "source": "index.js",
  "version": "1.0.0"
}
