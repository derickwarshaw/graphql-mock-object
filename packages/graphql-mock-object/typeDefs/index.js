export const MockAddress = `
  type MockAddress {
    city: String!
    cityPrefix: String!
    country: String!
    countryCode: String!
    county: String!
    latitude: Float!
    longitude: Float!
    secondaryAddress: String!
    state: String!
    stateAbbr: String!
    streetAddress: String!
    streetName: String!
    streetPrefix: String!
    streetSuffix: String!
    zipCode: String!
  }
`

export const MockCommerce = `
  type MockCommerce {
    color: String!
    department: String!
    price: Float!
    product: String!
    productAdjective: String!
    productName: String!
    productMaterial: String!
  }
`

export const MockCompany = `
  type MockCompany {
    bs: String!
    bsAdjective: String!
    bsBuzz: String!
    bsNoun: String!
    catchPhraseAdjective: String!
    catchPhraseDescriptor: String!
    catchPhrase: String!
    catchPhraseNoun: String!
    companyName: String!
    companySuffix: String!
    suffixes: [String!]!
  }
`

export const MockObject = `
  type MockObject {
    address: MockAddress!
    commerce: MockCommerce!
    company: MockCompany!
    Boolean: Boolean!
    Float: Float!
    ID(value: String): ID!
    Int: Int!
    List(length: Int = 0): [MockObject]!
    Mock(seed: Int): MockObject!
    Null: String
    String(value: String): String!
  }
`

export const Query = `
  type Query {
    Mock(seed: Int): MockObject
  }
`

export const typeDefs = [
  MockAddress,
  MockCommerce,
  MockCompany,
  MockObject,
  Query,
]
