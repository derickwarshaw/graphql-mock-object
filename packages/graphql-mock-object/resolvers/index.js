import faker from "faker"

import { version } from "../package.json"

// Seed is based on current version number. The intent is that any major
// version change will yield different resultsl
const majorVersion = parseInt(version.match(/\d+/g).shift(), 10) || 1

const fakerResolver = template => (parent, args) => {
  const { value = template } = args

  return faker.fake(value)
}

// The resolvers
export const MockAddress = {
  city: fakerResolver("{{address.city}}"),
  cityPrefix: fakerResolver("{{address.cityPrefix}}"),
  country: fakerResolver("{{address.country}}"),
  countryCode: fakerResolver("{{address.countryCode}}"),
  county: fakerResolver("{{address.county}}"),
  latitude: fakerResolver("{{address.latitude}}"),
  longitude: fakerResolver("{{address.longitude}}"),
  secondaryAddress: fakerResolver("{{address.secondaryAddress}}"),
  state: fakerResolver("{{address.state}}"),
  stateAbbr: fakerResolver("{{address.stateAbbr}}"),
  streetAddress: fakerResolver("{{address.streetAddress}}"),
  streetName: fakerResolver("{{address.streetName}}"),
  streetPrefix: fakerResolver("{{address.streetPrefix}}"),
  streetSuffix: fakerResolver("{{address.streetSuffix}}"),
  zipCode: fakerResolver("{{address.zipCode}}"),
}

export const MockCommerce = {
  color: fakerResolver("{{commerce.color}}"),
  department: fakerResolver("{{commerce.department}}"),
  price: fakerResolver("{{commerce.price}}"),
  product: fakerResolver("{{commerce.product}}"),
  productAdjective: fakerResolver("{{commerce.productAdjective}}"),
  productName: fakerResolver("{{commerce.productName}}"),
  productMaterial: fakerResolver("{{commerce.productMaterial}}"),
}

export const MockCompany = {
  bs: fakerResolver("{{company.bs}}"),
  bsAdjective: fakerResolver("{{company.bsAdjective}}"),
  bsBuzz: fakerResolver("{{company.bsBuzz}}"),
  bsNoun: fakerResolver("{{company.bsNoun}}"),
  catchPhraseAdjective: fakerResolver("{{company.catchPhraseAdjective}}"),
  catchPhraseDescriptor: fakerResolver("{{company.catchPhraseDescriptor}}"),
  catchPhrase: fakerResolver("{{company.catchPhrase}}"),
  catchPhraseNoun: fakerResolver("{{company.catchPhraseNoun}}"),
  companyName: fakerResolver("{{company.companyName}}"),
  companySuffix: fakerResolver("{{company.companySuffix}}"),
  suffixes(parent, args) {
    return faker.company.suffixes()
  },
}

export const MockDatabase = {
  collation: fakerResolver("{{database.collation}}"),
  column: fakerResolver("{{database.column}}"),
  engine: fakerResolver("{{database.engine}}"),
  type: fakerResolver("{{database.type}}"),
}

export const MockObject = {
  address: () => ({}),
  Boolean: fakerResolver("{{random.boolean}}"),
  commerce: () => ({}),
  company: () => ({}),
  database: () => ({}),
  Float: fakerResolver("0.{{random.number}}"),
  ID: fakerResolver("{{random.number}}"),
  Int: fakerResolver("{{random.number}}"),
  List(parent, args) {
    const { length } = args

    return Array.from({ length }).map(() => ({}))
  },
  Mock(parent, args) {
    const { seed = majorVersion } = args

    faker.seed(seed)

    return {}
  },
  String: fakerResolver("{{lorem.sentence}}"),
}

export const Query = {
  Mock: MockObject.Mock,
}

export const resolvers = {
  MockAddress,
  MockCommerce,
  MockCompany,
  MockDatabase,
  MockObject,
  Query,
}
