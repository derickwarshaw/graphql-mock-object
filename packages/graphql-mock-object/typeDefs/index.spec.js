import { runQuery } from "graphql-server-core"
import { inspect } from "util"

import { mockSchema as schema } from "../../graphql-mock-object"

describe("typeDefs", () => {
  describe("Query", () => {
    it("__type should match snapshot", async () => {
      const { data, errors } = await runQuery({
        query: `{
          Query: __type(name: "Query") {
            fields {
              name
              description
              args { name description defaultValue }
              type { name }
            }
          }
        }`,
        schema,
      })

      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
  })

  describe("MockObject", () => {
    it("__type should match snapshot", async () => {
      const { data, errors } = await runQuery({
        query: `{
          MockObject: __type(name: "MockObject") {
            inputFields { name description defaultValue }
            fields {
              name
              description
              args { name description defaultValue }
              type { name }
            }
          }
        }`,
        schema,
      })

      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })

    it("should return consistent, random values", async () => {
      const query = `{
        Mock {
          Boolean
          Float
          ID
          Int
          List(length: 2) {
            ID
          }
          Mock {
            ID
          }
          Null
          String
        }
      }`

      const { data, errors } = await runQuery({ query, schema })

      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()

      const duplicate = await runQuery({ query, schema })

      expect(duplicate.data).toEqual(data)
    })

    it("should support faker.js", async () => {
      const query = `{
        Mock {
          address {
            city
            cityPrefix
            country
            countryCode
            county
            latitude
            longitude
            secondaryAddress
            state
            stateAbbr
            streetAddress
            streetName
            streetPrefix
            streetSuffix
            zipCode
          }

          commerce {
            color
            department
            price
            product
            productAdjective
            productMaterial
            productName
          }

          company {
            bs
            bsAdjective
            bsBuzz
            bsNoun
            catchPhraseAdjective
            catchPhraseDescriptor
            catchPhrase
            catchPhraseNoun
            companyName
            companySuffix
            suffixes
          }

          database {
            collation
            column
            engine
            type
          }
        }
      }`

      const { data, errors } = await runQuery({ query, schema })

      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
  })
})
