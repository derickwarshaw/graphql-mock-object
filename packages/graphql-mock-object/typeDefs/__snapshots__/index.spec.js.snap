// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeDefs MockObject __type should match snapshot 1`] = `
Object {
  "MockObject": Object {
    "fields": Array [
      Object {
        "args": Array [],
        "description": "",
        "name": "address",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "Boolean",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "commerce",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "company",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "database",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "date",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "finance",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "Float",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "hacker",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": null,
            "description": "",
            "name": "value",
          },
        ],
        "description": "",
        "name": "ID",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "image",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "Int",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "internet",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": "0",
            "description": "",
            "name": "length",
          },
        ],
        "description": "",
        "name": "List",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "lorem",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": null,
            "description": "",
            "name": "seed",
          },
        ],
        "description": "",
        "name": "Mock",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "name",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "Null",
        "type": Object {
          "name": "String",
        },
      },
      Object {
        "args": Array [],
        "description": "",
        "name": "phone",
        "type": Object {
          "name": null,
        },
      },
      Object {
        "args": Array [
          Object {
            "defaultValue": null,
            "description": "",
            "name": "value",
          },
        ],
        "description": "",
        "name": "String",
        "type": Object {
          "name": null,
        },
      },
    ],
    "inputFields": null,
  },
}
`;

exports[`typeDefs MockObject should return consistent, random values 1`] = `
Object {
  "Mock": Object {
    "Boolean": true,
    "Float": 0.7072,
    "ID": "70814",
    "Int": 83994,
    "List": Array [
      Object {
        "ID": "29090",
      },
      Object {
        "ID": "12132",
      },
    ],
    "Mock": Object {
      "ID": "55079",
    },
    "Null": null,
    "String": "Impedit voluptas nostrum.",
  },
}
`;

exports[`typeDefs MockObject should support faker.js 1`] = `
Object {
  "Mock": Object {
    "address": Object {
      "city": "Antonechester",
      "cityPrefix": "Lake",
      "country": "Finland",
      "countryCode": "BQ",
      "county": "Borders",
      "latitude": 12.476,
      "longitude": 141.4609,
      "secondaryAddress": "Apt. 801",
      "state": "Arizona",
      "stateAbbr": "HI",
      "streetAddress": "00460 Klein Crossroad",
      "streetName": "Gottlieb Ports",
      "streetPrefix": "b",
      "streetSuffix": "Ranch",
      "zipCode": "59075",
    },
    "commerce": Object {
      "color": "silver",
      "department": "Computers",
      "price": 402,
      "product": "Shirt",
      "productAdjective": "Awesome",
      "productMaterial": "Plastic",
      "productName": "Incredible Metal Soap",
    },
    "company": Object {
      "bs": "magnetic morph platforms",
      "bsAdjective": "bleeding-edge",
      "bsBuzz": "deliver",
      "bsNoun": "deliverables",
      "catchPhrase": "Front-line demand-driven contingency",
      "catchPhraseAdjective": "Seamless",
      "catchPhraseDescriptor": "secondary",
      "catchPhraseNoun": "success",
      "companyName": "Raynor - Weber",
      "companySuffix": "Inc",
      "suffixes": Array [
        "Inc",
        "and Sons",
        "LLC",
        "Group",
      ],
    },
    "database": Object {
      "collation": "cp1250_general_ci",
      "column": "title",
      "engine": "BLACKHOLE",
      "type": "timestamp",
    },
    "date": Object {
      "between": "Mon Jan 01 2018 00:00:00 GMT-0800 (PST)",
      "future": "Tue Oct 23 2018 07:13:11 GMT-0700 (PDT)",
      "month": "November",
      "past": "Wed May 30 2018 16:03:32 GMT-0700 (PDT)",
      "recent": "Wed May 30 2018 06:59:12 GMT-0700 (PDT)",
      "weekday": "Tuesday",
    },
    "finance": Object {
      "account": 4675730,
      "accountName": "Savings Account",
      "amount": 315.77,
      "bic": "OMTEMQH1",
      "bitcoinAddress": "1NUNLVATADG27R72IU3OHGDDPIQG",
      "currencyCode": "SAR",
      "currencyName": "Trinidad and Tobago Dollar",
      "currencySymbol": "ï·¼",
      "iban": "DE30845851030092005075",
      "mask": "4952",
      "transactionType": "deposit",
    },
    "hacker": Object {
      "abbreviation": "JSON",
      "adjective": "primary",
      "ingverb": "compressing",
      "noun": "microchip",
      "phrase": "We need to synthesize the mobile GB bandwidth!",
      "verb": "hack",
    },
    "image": Object {
      "abstract": "http://lorempixel.com/640/480/abstract",
      "animals": "http://lorempixel.com/640/480/animals",
      "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/juanmamartinez/128.jpg",
      "business": "http://lorempixel.com/640/480/business",
      "cats": "http://lorempixel.com/640/480/cats",
      "city": "http://lorempixel.com/640/480/city",
      "dataUri": "data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20version%3D%221.1%22%20baseProfile%3D%22full%22%20width%3D%22undefined%22%20height%3D%22undefined%22%3E%20%3Crect%20width%3D%22100%25%22%20height%3D%22100%25%22%20fill%3D%22grey%22%2F%3E%20%20%3Ctext%20x%3D%220%22%20y%3D%2220%22%20font-size%3D%2220%22%20text-anchor%3D%22start%22%20fill%3D%22white%22%3Eundefinedxundefined%3C%2Ftext%3E%20%3C%2Fsvg%3E",
      "fashion": "http://lorempixel.com/640/480/fashion",
      "food": "http://lorempixel.com/640/480/food",
      "image": "http://lorempixel.com/640/480/fashion",
      "imageUrl": "http://lorempixel.com/640/480",
      "nature": "http://lorempixel.com/640/480/nature",
      "nightlife": "http://lorempixel.com/640/480/nightlife",
      "people": "http://lorempixel.com/640/480/people",
      "sports": "http://lorempixel.com/640/480/sports",
      "technics": "http://lorempixel.com/640/480/technics",
      "transport": "http://lorempixel.com/640/480/transport",
    },
    "internet": Object {
      "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/sgaurav_baghel/128.jpg",
      "color": "#180b7a",
      "domainName": "malvina.net",
      "domainSuffix": "biz",
      "domainWord": "kamryn",
      "email": "Melany42@yahoo.com",
      "exampleEmail": "Reese_Smitham@example.net",
      "ip": "225.122.204.215",
      "ipv6": "5bca:eeca:45e8:83d3:3b3c:ffcd:3871:872e",
      "mac": "4c:47:e7:83:92:51",
      "password": "ZuH94f8vNpBJTOr",
      "protocol": "https",
      "url": "http://kamryn.net",
      "userName": "Rodolfo99",
    },
    "lorem": Object {
      "lines": "Eligendi rem saepe impedit.
Laboriosam et optio et placeat.",
      "paragraph": "Quod pariatur nesciunt vitae optio nemo voluptas natus qui. Ea explicabo est voluptate eos ut ipsam est quam enim. Voluptas sequi dolores aut nihil quibusdam soluta ut fuga consequatur. Quia nesciunt labore qui eum corporis delectus odit cumque quis.",
      "paragraphs": "Quis labore fugiat optio officiis accusantium sunt non. Commodi fugiat et pariatur. Ipsam id minima incidunt et dignissimos ut facilis. Dolor quo autem.
 
Est voluptate incidunt odit delectus odio. At nisi ut hic quisquam cumque velit voluptas eos. Voluptatem illo modi. Ex corporis dolor sint ducimus sequi voluptatum reiciendis. Quas consequatur eum voluptatem architecto et deleniti.
 
Ea rerum rerum. Nemo quasi veniam beatae numquam ut. Architecto consequuntur sit asperiores iusto et. Et qui reiciendis dolorum exercitationem quis. Dolorum in nesciunt temporibus ad et. Deserunt sequi consequuntur modi doloremque et incidunt deleniti perferendis.",
      "sentence": "Ad doloremque earum.",
      "sentences": "Neque architecto soluta ad quibusdam. In nihil eligendi qui reprehenderit et quae sit odio maxime. Consequatur mollitia omnis dolores autem. Harum vel omnis voluptatem.",
      "slug": "quidem-velit-expedita",
      "text": "Provident ratione dolor cumque assumenda.",
      "word": "similique",
      "words": "accusantium laborum et",
    },
    "name": Object {
      "findName": "Mr. Haylee McCullough",
      "firstName": "Barney",
      "jobArea": "Functionality",
      "jobDescriptor": "Global",
      "jobTitle": "International Data Associate",
      "jobType": "Associate",
      "lastName": "Greenholt",
      "prefix": "Dr.",
      "suffix": "DVM",
      "title": "International Usability Liaison",
    },
    "phone": Object {
      "phoneFormats": "(###) ###-#### x###",
      "phoneNumber": "1-786-649-8318 x840",
      "phoneNumberFormat": "887-614-7684",
    },
  },
}
`;

exports[`typeDefs Query __type should match snapshot 1`] = `
Object {
  "Query": Object {
    "fields": Array [
      Object {
        "args": Array [
          Object {
            "defaultValue": null,
            "description": "",
            "name": "seed",
          },
        ],
        "description": "",
        "name": "Mock",
        "type": Object {
          "name": "MockObject",
        },
      },
    ],
  },
}
`;
